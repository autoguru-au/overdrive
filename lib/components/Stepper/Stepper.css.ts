import { style, styleVariants } from '@vanilla-extract/css';

import { overdriveTokens as vars } from '../../themes/theme.css';

export const disabled = style({ cursor: 'not-allowed' });

export const root = style({
	width: 'max-content',
	selectors: {
		[`&:not(${disabled}):focus`]: {
			borderColor: vars.colours.intent.information.background.standard,
		},
	},
});

export const width = styleVariants({
	full: {
		width: '100%',
	},
	default: {
		width: 'max-content',
	},
});
export const handle = styleVariants({
	default: {
		width: vars.space[6],
		height: vars.space[6],
		transition: `background-color 0.1s ${vars.animation.easing.standard}`,
		selectors: {
			[`${root}:not(${disabled}) &:hover`]: {
				backgroundColor: vars.colours.intent.primary.background.strong,
			},
			[`${root}:not(${disabled}) &:active`]: {
				backgroundColor: vars.colours.intent.primary.background.strong,
			},
		},
	},
	disabled: {
		cursor: 'not-allowed',
	},
});

export const label = style({
	minWidth: vars.space['8'],
});
