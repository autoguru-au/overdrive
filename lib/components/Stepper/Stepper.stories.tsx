import { action } from '@storybook/addon-actions';
import type { Meta, StoryObj } from '@storybook/react';
import React, { useState } from 'react';

import { Stepper } from '.';

const meta = {
	title: 'Forms & Input Fields/Stepper',
	component: Stepper,
} satisfies Meta<typeof Stepper>;

export default meta;
type Story = StoryObj<typeof Stepper>;

const Template: Story = {
	render: ({ onChange: incomingOnChange, value: incomingValue, ...args }) => {
		const [value, setValue] = useState<number>(incomingValue!);
		return (
			<Stepper
				onChange={(stepValue) => {
					setValue(stepValue);
					if (incomingOnChange) incomingOnChange(stepValue);
				}}
				value={value}
				{...args}
			/>
		);
	},
};

export const Standard: Story = {
	...Template,
	args: {
		value: 2.5,
		min: 0,
		max: 10,
		step: 0.5,
		disabled: false,
		onChange: action('onChange'),
	},
};

export const FullWidth: Story = {
	...Template,
	args: {
		...Standard.args,
		isFullWidth: true,
	},
};

const formatter = new Intl.NumberFormat('en-AU', {
	style: 'currency',
	currency: 'AUD',
});

export const WithFormatting: Story = {
	...Template,
	args: {
		value: 32,
		min: 0,
		max: 99,
		step: 0.05,
		disabled: false,
		format: formatter.format,
		onChange: action('onChange'),
	},
};
