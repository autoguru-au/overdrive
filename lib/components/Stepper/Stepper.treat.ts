import { style, styleMap } from 'treat';

export const flexCenter = style({
	display: 'flex',
	alignContent: 'center',
	alignItems: 'center',
	flexDirection: 'row',
	justifyContent: 'center',
});

export const disabled = style({ cursor: 'not-allowed' });

export const root = style(theme => ({
	userSelect: 'none',
	color: theme.colours.gamut.gray900,
	outline: 'none',

	selectors: {
		[`&:not(${disabled}):focus`]: {
			borderColor: theme.colours.gamut.blue500,
		},
	},
}));

export const handle = styleMap(theme => ({
	default: {
		width: theme.space[6],
		height: theme.space[6],
		cursor: 'pointer',
		color: theme.colours.gamut.white,
		outline: 'none',
		backgroundColor: theme.colours.gamut.green600,
		transition: `background-color 0.1s ${theme.animation.easing.standard}`,

		selectors: {
			[`${root}:not(${disabled}) &:hover`]: {
				backgroundColor: theme.colours.gamut.green700,
			},
			[`${root}:not(${disabled}) &:active`]: {
				backgroundColor: theme.colours.gamut.green800,
			},
		},
	},
	disabled: {
		borderColor: theme.colours.gamut.green200,
		backgroundColor: theme.colours.gamut.green200,
	},
}));

export const label = style(theme => ({
	flex: '1 0 auto',
	minWidth: theme.space['8'],
	padding: `0 ${theme.space['2']}`,
	textAlign: 'center',
}));
