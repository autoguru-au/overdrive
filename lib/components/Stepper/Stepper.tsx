import { clamp } from '@autoguru/utilities';
import clsx from 'clsx';
import React, {
	FunctionComponent,
	memo,
	useCallback,
	useReducer,
	useRef,
} from 'react';
import { Icon, MinusIcon, PlusIcon } from '../Icon';
import styles from './style.scss';

export interface IProps {
	className?: string;
	disabled?: boolean;
	value?: number;
	min?: number;
	max?: number;
	step?: number;

	format?(value: number): string;

	onChange?(value: number): void;
}

enum EActionType {
	INCREMENT,
	DECREMENT,
	VALUE,
}

interface IAction {
	type: EActionType;
	value?: number;
	min: number;
	max: number;
	step: number;
}

interface IState {
	value: number;
}

const reducer = (state: IState, action: IAction) => {
	switch (action.type) {
		case EActionType.INCREMENT:
			return {
				value: clamp(state.value + action.step, action.min, action.max),
			};
		case EActionType.DECREMENT:
			return {
				value: clamp(state.value - action.step, action.min, action.max),
			};
		case EActionType.VALUE:
			return { value: action.value };

		/* istanbul ignore next */
		default:
			throw new Error();
	}
};

const StepperComponent: FunctionComponent<IProps> = ({
	className = '',
	disabled = false,
	step = 1,
	min = Number.NEGATIVE_INFINITY,
	max = Number.POSITIVE_INFINITY,
	value,
	format = value => value.toString(),
	onChange,
}) => {
	const isDisabled: boolean = disabled || value === void 0 || value === null;

	const [state, dispatch] = useReducer(reducer, { value });

	const prevValue = useRef(value);

	if (state.value !== value && typeof onChange === 'function') {
		onChange(state.value);
	}

	const onDecrement = useCallback(
		() =>
			dispatch({
				type: EActionType.DECREMENT,
				step,
				min,
				max,
			}),
		[step, min, max]
	);
	const onIncrement = useCallback(
		() =>
			dispatch({
				type: EActionType.INCREMENT,
				step,
				min,
				max,
			}),
		[step, min, max]
	);

	if (prevValue.current !== value) {
		dispatch({
			type: EActionType.VALUE,
			value,
			step,
			min,
			max,
		});
		prevValue.current = value;
	}

	return (
		<div
			className={clsx([styles.root, className])}
			aria-disabled={isDisabled}>
			<button
				className={styles.handle}
				onClick={onDecrement}
				aria-label="step down"
				aria-disabled={isDisabled}
				disabled={isDisabled}>
				<Icon icon={MinusIcon} size={16} />
			</button>
			<span className={styles.label}>
				{Number.isFinite(state.value) ? format(state.value) : ''}
			</span>
			<button
				className={styles.handle}
				onClick={onIncrement}
				aria-label="step up"
				aria-disabled={isDisabled}
				disabled={isDisabled}>
				<Icon icon={PlusIcon} size={16} />
			</button>
		</div>
	);
};

export const Stepper = memo(StepperComponent);
