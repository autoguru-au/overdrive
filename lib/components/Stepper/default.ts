export { Stepper as default } from './Stepper';
