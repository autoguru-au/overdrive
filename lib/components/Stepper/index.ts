export { Stepper, IProps as IStepperProps } from './Stepper';
