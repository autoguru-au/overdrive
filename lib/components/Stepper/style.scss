@import '../../theme/tokens/facade';

@mixin isDisabledState() {
	&[disabled],
	&[aria-disabled='true'] {
		cursor: not-allowed;

		&:not(.loading) {
			@content;
		}
	}
}

$handleSize: 24px;

.root,
.handle {
	flex-direction: row;
	display: flex;
	align-content: center;
	justify-content: center;
	align-items: center;
}

.root {
	font-size: var(--global--fontSize--5);
	line-height: var(--global--fontSize--5);
	color: getColour(gray, 900);
	border-radius: 4px;
	border: solid 1px getColour(gray, 300);
	box-shadow: 0 1px 10px 0 rgba(0, 0, 0, 0.03),
		0 4px 5px 0 rgba(0, 0, 0, 0.03), 0 2px 4px -1px rgba(0, 0, 0, 0.05);
	padding: calc(var(--global--space--3) - 1px)
		calc(var(--global--space--2) - 1px);

	@include isDisabledState() {
		cursor: not-allowed;
	}
}

.handle {
	width: $handleSize;
	height: $handleSize;
	background-color: getColour(green, 700);
	border-radius: 50%;
	cursor: pointer;
	outline: none;
	fill: white;
	padding: 0;

	@include isDisabledState() {
		background-color: getColour(green, 200);
		border-color: getColour(green, 200);
	}
}

.label {
	min-width: 64px;
	text-align: center;
	padding: 0 var(--global--space--2);
	flex: 1 0 auto;
}
