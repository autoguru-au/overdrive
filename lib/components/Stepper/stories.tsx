import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { boolean, number } from '@storybook/addon-knobs';
import React, { useState } from 'react';
import { Stepper } from './Stepper';

const badgeProps = () => ({
	min: number('min', 0),
	max: number('max', 10),
	step: number('step', 0.5),
	disabled: boolean('disabled', false),
});

storiesOf('Components|Inputs/Stepper', module).add('default', () => {
	const StepperWrapper = () => {
		const [value, setValue] = useState(1);

		const onChangeHandler = changedValue => {
			action('onChange')(changedValue);
			setValue(changedValue);
		};

		const formatFn = input => `$${input}`;

		return (
			<Stepper
				{...badgeProps()}
				format={formatFn}
				value={number('value', value)}
				onChange={onChangeHandler}
			/>
		);
	};

	return <StepperWrapper />;
});
