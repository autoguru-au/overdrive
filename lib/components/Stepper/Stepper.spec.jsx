import React, { useState } from 'react';
import { mount, render, shallow } from 'enzyme';
import { Stepper } from './Stepper';
import { act } from 'react-dom/test-utils';

describe('<Stepper />', () => {
	it('should not throw', () =>
		expect(() => shallow(<Stepper />)).not.toThrow());

	it('should match snapshot without props', () => {
		expect(render(<Stepper />)).toMatchSnapshot();
	});

	it('should match snapshot with props', () => {
		expect(
			render(<Stepper min={-20} max={20} value={19} step={1.5} />),
		).toMatchSnapshot();
	});

	it('should match snapshot with label formatter', () => {
		expect(
			render(
				<Stepper
					min={0}
					max={100}
					value={25}
					step={5}
					format={value => `${value}%`}
				/>,
			),
		).toMatchSnapshot();
	});

	it('should pass on className to dom element', () => {
		const stepper = mount(<Stepper className="stepper-class" value={10} />);
		expect(stepper.find('div.stepper-class').length === 1).toBeTruthy();
	});

	it('should add a span element with the formatted value in it', () => {
		const stepper = mount(
			<Stepper
				min={0}
				max={100}
				value={99}
				step={0.5}
				format={value => `$${value}`}
			/>,
		);
		expect(
			stepper
				.find('span')
				.first()
				.text(),
		).toEqual('$99');
	});

	it('should display passed down value even when outside acceptable boundaries', () => {
		const stepper = mount(
			<Stepper
				min={-100}
				max={40}
				value={45}
				step={1}
				format={value => `${value}%`}
			/>,
		);
		expect(
			stepper
				.find('span')
				.first()
				.text(),
		).toEqual('45%');
	});

	it('should fire change with the correct changed value after next and previous button click', () => {
		const spyedCallback = jest.fn(value => `${value}%`);

		const wrapper = mount(
			<Stepper
				min={-100}
				max={100}
				value={45}
				step={5}
				format={spyedCallback}
			/>,
		);

		expect(
			wrapper
				.find('span')
				.first()
				.text(),
		).toEqual('45%');

		wrapper
			.find('button')
			.at(1)
			.simulate('click');

		expect(spyedCallback).toHaveBeenCalledWith(50);

		wrapper
			.find('button')
			.at(1)
			.simulate('click');

		expect(spyedCallback).toHaveBeenCalledWith(55);

		wrapper
			.find('button')
			.at(0)
			.simulate('click');

		expect(spyedCallback).toHaveBeenCalledWith(50);

		wrapper.unmount();
	});

	it('should update its value when and a value prop comes in', done => {
		const spyedCallback = jest.fn(value => `$${value}`);
		const StepperWrapper = ({ getValueSetter }) => {
			const [value, setValue] = useState(16);

			getValueSetter(setValue);

			return (
				<Stepper
					min={-100}
					max={100}
					value={value}
					step={0.5}
					format={spyedCallback}
				/>
			);
		};

		const wrapper = mount(<StepperWrapper getValueSetter={onSetValue} />);

		function onSetValue(setValue) {
			setTimeout(() => {
				expect(
					wrapper
						.find('span')
						.first()
						.text(),
				).toEqual('$16');

				act(() => {
					setValue(199);
				});

				expect(
					wrapper
						.find('span')
						.first()
						.text(),
				).toEqual('$199');

				wrapper
					.find('button')
					.at(1)
					.simulate('click');

				expect(
					wrapper
						.find('span')
						.first()
						.text(),
				).toEqual('$100');

				act(() => {
					setValue(250);
				});

				wrapper
					.find('button')
					.at(1)
					.simulate('click');

				expect(
					wrapper
						.find('span')
						.first()
						.text(),
				).toEqual('$100');

				expect(spyedCallback).toHaveBeenCalledTimes(7);

				wrapper.unmount();
				done();
			}, 100);
		}
	});

	it('should onChange when the value changes', () => {
		const onChangeHandler = jest.fn();

		const stepper = mount(
			<Stepper
				min={-100}
				max={100}
				step={0.5}
				value={1}
				onChange={onChangeHandler}
			/>,
		);

		expect(
			stepper
				.find('span')
				.first()
				.text(),
		).toEqual('1');

		stepper
			.find('button')
			.at(1)
			.simulate('click');

		expect(onChangeHandler).toHaveBeenCalledTimes(1);
		expect(onChangeHandler).toHaveBeenCalledWith(1.5);

		stepper
			.find('button')
			.at(1)
			.simulate('click');

		expect(onChangeHandler).toHaveBeenCalledTimes(2);
		expect(onChangeHandler).toHaveBeenCalledWith(2);
	});
});
