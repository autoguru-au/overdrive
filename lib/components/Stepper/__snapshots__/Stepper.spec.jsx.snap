// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stepper /> should match snapshot with label formatter 1`] = `
<div
  aria-disabled="false"
  class="base 3_mobile_base 3_mobile_base 3_mobile_base 3_mobile_base border_style border_colour_top_gray_base border_colour_right_gray_base border_colour_bottom_gray_base border_colour_left_gray_base 1_mobile_base 1_mobile_base 1_mobile_base 1_mobile_base 2_mobile_base 1_mobile_base root_base flexCenter"
  tabindex="0"
>
  <button
    aria-label="step down"
    class="base appearance cursorPointer button none_mobile_base none_mobile_base none_mobile_base none_mobile_base full_mobile_base handle_default_base flexCenter"
    tabindex="-1"
  >
    <i
      class="base display_block small_mobile_base"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 12.998H5V11l14-.002v2z"
        />
      </svg>
    </i>
  </button>
  <span
    class="base root_base sizes_4_base align_left colours_neutral_base fontWeight_normal_base label_base"
  >
    25%
  </span>
  <button
    aria-label="step up"
    class="base appearance cursorPointer button none_mobile_base none_mobile_base none_mobile_base none_mobile_base full_mobile_base handle_default_base flexCenter"
    tabindex="-1"
  >
    <i
      class="base display_block small_mobile_base"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 12.998h-6v6h-2v-6H5v-2h6v-6h2v6h6v2z"
        />
      </svg>
    </i>
  </button>
</div>
`;

exports[`<Stepper /> should match snapshot with props 1`] = `
<div
  aria-disabled="false"
  class="base 3_mobile_base 3_mobile_base 3_mobile_base 3_mobile_base border_style border_colour_top_gray_base border_colour_right_gray_base border_colour_bottom_gray_base border_colour_left_gray_base 1_mobile_base 1_mobile_base 1_mobile_base 1_mobile_base 2_mobile_base 1_mobile_base root_base flexCenter"
  tabindex="0"
>
  <button
    aria-label="step down"
    class="base appearance cursorPointer button none_mobile_base none_mobile_base none_mobile_base none_mobile_base full_mobile_base handle_default_base flexCenter"
    tabindex="-1"
  >
    <i
      class="base display_block small_mobile_base"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 12.998H5V11l14-.002v2z"
        />
      </svg>
    </i>
  </button>
  <span
    class="base root_base sizes_4_base align_left colours_neutral_base fontWeight_normal_base label_base"
  >
    19
  </span>
  <button
    aria-label="step up"
    class="base appearance cursorPointer button none_mobile_base none_mobile_base none_mobile_base none_mobile_base full_mobile_base handle_default_base flexCenter"
    tabindex="-1"
  >
    <i
      class="base display_block small_mobile_base"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 12.998h-6v6h-2v-6H5v-2h6v-6h2v6h6v2z"
        />
      </svg>
    </i>
  </button>
</div>
`;

exports[`<Stepper /> should match snapshot without props 1`] = `
<div
  aria-disabled="true"
  class="base 3_mobile_base 3_mobile_base 3_mobile_base 3_mobile_base border_style border_colour_top_gray_base border_colour_right_gray_base border_colour_bottom_gray_base border_colour_left_gray_base 1_mobile_base 1_mobile_base 1_mobile_base 1_mobile_base 2_mobile_base 1_mobile_base root_base flexCenter"
  tabindex="0"
>
  <button
    aria-label="step down"
    class="base appearance cursorPointer button none_mobile_base none_mobile_base none_mobile_base none_mobile_base full_mobile_base handle_default_base flexCenter"
    disabled=""
    tabindex="-1"
  >
    <i
      class="base display_block small_mobile_base"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 12.998H5V11l14-.002v2z"
        />
      </svg>
    </i>
  </button>
  <span
    class="base root_base sizes_4_base align_left colours_neutral_base fontWeight_normal_base label_base"
  >
    0
  </span>
  <button
    aria-label="step up"
    class="base appearance cursorPointer button none_mobile_base none_mobile_base none_mobile_base none_mobile_base full_mobile_base handle_default_base flexCenter"
    disabled=""
    tabindex="-1"
  >
    <i
      class="base display_block small_mobile_base"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 12.998h-6v6h-2v-6H5v-2h6v-6h2v6h6v2z"
        />
      </svg>
    </i>
  </button>
</div>
`;
